### PR TITLE
control: init OSQP solver for MPC

### PR DESCRIPTION
**Pull request by sjiang2018**

### CODE REVIEW

File: modules/common/math/mpc_osqp.cc [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+          state_total_dim + i * control_dim_ + j, matrix_r_(j, j));
+      ++value_index;
+    }
+  }
+  CHECK_EQ(value_index, num_param_);
+
+  int ind_p = 0;
+  for (int i = 0; i < num_param_; ++i) {
+    // TODO(SHU) Check this
+    P_indptr->push_back(ind_p);
```
Comment (jinghaomiao) [inline modules/common/math/mpc_osqp.cc]: Use emplace_back as much as possible.

File: modules/common/math/mpc_osqp.cc [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+      Eigen::MatrixXd::Identity(num_param_, num_param_);
+
+  matrix_constraint.block(state_dim_ * (horizon_ + 1), 0, num_param_,
+                          num_param_) = all_identity_mat;
+
+  std::vector<std::vector<std::pair<c_int, c_float>>> columns;
+  columns.resize(num_param_ + 1);  // TODO(SHU): double check
+  int value_index = 0;
+  // state and terminal state
+  for (int i = 0; i < num_param_; ++i) {                                // col
```
Comment (jinghaomiao) [inline modules/common/math/mpc_osqp.cc]: Use size_t when applicable to avoid out of boundary violation for vectors.

File: modules/common/math/mpc_osqp.cc (hunk 1 above)
Comment (sjiang2018) [reply modules/common/math/mpc_osqp.cc]: Fixed. Thank you.

File: modules/common/math/mpc_osqp.cc (hunk 2 above)
Comment (sjiang2018) [reply modules/common/math/mpc_osqp.cc]: Fixed. Thank you.

File: modules/control/controller/mpc_controller.cc [hunk 3]
```diff
@@ -372,41 +373,57 @@ Status MPCController::ComputeControlCommand(
...
+
+  std::vector<double> *control_cmd;
+  if (FLAGS_use_osqp_solver) {
+    apollo::common::math::MpcOsqp mpc_osqp(
+        matrix_ad_, matrix_bd_, matrix_q_updated_, matrix_r_updated_,
+        lower_bound, upper_bound, matrix_state_, mpc_max_iteration_);
+    if (!mpc_osqp.MpcOsqpSolver(control_cmd)) {
+      AERROR << "MPC OSQP solver failed";
+    } else {
+      AINFO << "MPC OSQP problem solved! ";
```
Comment (jinghaomiao) [inline modules/control/controller/mpc_controller.cc]: Use ADEBUG instead of AINFO to reduce unnecessary logging.

Comment (sjiang2018) [reply modules/control/controller/mpc_controller.cc]: Fixed. Thank you!

File: modules/common/math/mpc_osqp.cc [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+  for (size_t i = 0; i < control_dim_; ++i) {
+    control_cmd->at(i) = osqp_workspace->solution->x[i + first_control];
+  }
+
+  // Cleanup
+  osqp_cleanup(osqp_workspace);
+  FreeData(data);
+  c_free(settings);
+  return true;
+}  // namespace math
```
Comment (HongyiSun) [inline modules/common/math/mpc_osqp.cc]: Duplicate namespace

File: modules/common/math/mpc_osqp_test.cc [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  upper_bound << 10, 10;
+
+  Eigen::MatrixXd initial_state(states, 1);
+  initial_state << 20, 0, 0, 0;
+  std::vector<double> control_cmd(2, 0);
+  ADEBUG << "IN TEST initialize";
+  MpcOsqp mpc_osqp_solver(A, B, Q, R, lower_bound, upper_bound, initial_state,
+                          max_iter, horizon);
+  ADEBUG << "IN TEST SOLVE";
+  mpc_osqp_solver.MpcOsqpSolver(&control_cmd);
```
Comment (Capri2014) [inline modules/common/math/mpc_osqp_test.cc]: Is it better to call the api 'solve', or 'get_results'?

File: modules/common/math/mpc_osqp_test.cc [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+#include "modules/common/math/mpc_osqp.h"
+
+#include "cyber/common/log.h"
+#include "gtest/gtest.h"
+
+namespace apollo {
+namespace common {
+namespace math {
+
+TEST(MPCOSQPSolverTest, MPCOSQP) {
```
Comment (Capri2014) [inline modules/common/math/mpc_osqp_test.cc]: Add test against null matrix and again non-full rank matrix

File: modules/common/math/mpc_osqp_test.cc (hunk 5 above)
Comment (sjiang2018) [reply modules/common/math/mpc_osqp_test.cc]: Fixed. Thank you.

File: modules/common/math/mpc_osqp_test.cc (hunk 6 above)
Comment (sjiang2018) [reply modules/common/math/mpc_osqp_test.cc]: Added. Thank you.
